text/src/autogluon/text/automm/predictor.py

```diff
@@ -578,91 +756,127 @@ def _fit(
+            if config.optimization.top_k_average_method == GREEDY_SOUP:
+                # Select the ingredients based on the methods proposed in paper
+                #  "Model soups: averaging weights of multiple fine-tuned models improves accuracy without
+                #  increasing inference time", https://arxiv.org/pdf/2203.05482.pdf
+                monitor_op = {MIN: operator.le, MAX: operator.ge}[minmax_mode]
+
                 logger.info(
-                    f"Start to fuse {config.optimization.top_k} checkpoints via the uniform soup algorithm."
+                    f"Start to fuse {len(top_k_model_paths)} checkpoints via the greedy soup algorithm."
                 )
-                ingredients = all_state_dicts
-            else:
-                if not best_k_models_path:
-                    # In the case of ddp_spawn, the best_k_models_path can be empty. This is due the
-                    # limitation of PT Lightning: https://github.com/PyTorchLightning/pytorch-lightning/issues/5582
-                    # Thus, we will need to reevaluate the validation performance of these checkpoints
-                    candidate_performance = []
-                    for cand_state_dict in all_state_dicts:
-                        self._model = self._load_state_dict(
-                            model=model,
-                            state_dict=cand_state_dict,
-                        )
-                        candidate_performance.append(self.evaluate(val_df,
-                                                                   [validation_metric_name])[validation_metric_name])
-                    all_state_dicts = [v[0] for v in sorted(zip(all_state_dicts, candidate_performance),
-                                                            key=lambda ele: ele[1], reverse=(minmax_mode == MAX))]
-                if config.optimization.top_k_average_method == GREEDY_SOUP:
-                    # Select the ingredients based on the methods proposed in paper
-                    #  "Model soups: averaging weights of multiple fine-tuned models improves accuracy without
-                    #  increasing inference time", https://arxiv.org/pdf/2203.05482.pdf
-                    monitor_op = {MIN: operator.le, MAX: operator.ge}[minmax_mode]
-
-                    logger.info(
-                        f"Start to fuse {config.optimization.top_k} checkpoints via the greedy soup algorithm."
+
+                ingredients = [top_k_model_paths[0]]
+                self._model = self._load_state_dict(
+                    model=model,
+                    path=top_k_model_paths[0],
+                    prefix=prefix,
+                )
+                best_score = self.evaluate(val_df, [validation_metric_name])[validation_metric_name]
+                for i in range(1, len(top_k_model_paths)):
+                    cand_avg_state_dict = average_checkpoints(
+                        checkpoint_paths=ingredients + [top_k_model_paths[i]],
```

Comment:
I know it's not in the original paper, but should we try to evaluate all the pre-selected models here through multiple iterations (including already added) and add only the best one improving the soup? Considering models already selected in previous iterations would allow to dynamically adjust soup component weights (models added: A-B-C-A results in 0.5-0.25-0.25 weights for models A-B-C respectively).

Reply:
Make sense. We can compare the performance on our benchmark datasets. If doing in this way, the number of iterations may become a new hyper-parameter since it doesn't need to be `len(top_k_model_paths)`. If we use `while True` instead, we may fall into dead loops for the condition operators `operator.le` and `operator.ge`.